File: scripts/build_release.py
#!/usr/bin/env python3
"""Build skill and plugin ZIPs from an explicit source-file allowlist."""

import argparse
import json
import re
from pathlib import Path
from zipfile import ZIP_DEFLATED, ZipFile, ZipInfo

ROOT = Path(__file__).resolve().parents[1]
SKILL_NAME = "instagram-saved-ids"
SKILL_FILES = (
    "SKILL.md",
    "agents/openai.yaml",
    "references/browser-collection.md",
    "references/runtime.md",
    "scripts/export_ids.js",
)
PLUGIN_FILES = ("plugin.json", ".codex-plugin/plugin.json", "README.md")


def build(root, output):
    root, output = Path(root).resolve(), Path(output).resolve()
    manifest = json.loads((root / "plugin.json").read_text(encoding="utf-8"))
    name, version = manifest["name"], manifest["version"]
    if not re.fullmatch(r"[a-z0-9]+(?:-[a-z0-9]+)*", name):
        raise ValueError("Invalid package name")
    if not re.fullmatch(r"\d+\.\d+\.\d+", version):
        raise ValueError("Release version must use major.minor.patch")
    legacy = json.loads((root / ".codex-plugin/plugin.json").read_text(encoding="utf-8"))
    for field in ("name", "version", "description", "license", "repository"):
        if legacy.get(field) != manifest.get(field):
            raise ValueError(f"Compatibility manifest differs on {field}")

    skill_dir = root / "skills" / SKILL_NAME
    skill_entries = [(skill_dir / path, f"{SKILL_NAME}/{path}") for path in SKILL_FILES]
    plugin_entries = [(root / path, path) for path in PLUGIN_FILES]
    plugin_entries.extend((skill_dir / path, f"skills/{SKILL_NAME}/{path}") for path in SKILL_FILES)
    if (root / "LICENSE").is_file():
        skill_entries.append((root / "LICENSE", f"{SKILL_NAME}/LICENSE"))
        plugin_entries.append((root / "LICENSE", "LICENSE"))
    elif manifest.get("license"):
        raise ValueError("Manifest declares a license but LICENSE is missing")

    # Check all inputs before opening outputs, and prevent symlink escapes.
    for source, _ in skill_entries + plugin_entries:
        if not source.is_file() or root not in source.resolve().parents:
            raise ValueError(f"Missing or external package input: {source}")
    outputs = [
        (output / f"{SKILL_NAME}-{version}.zip", skill_entries),
        (output / f"{name}-{version}.zip", plugin_entries),
    ]
    output.mkdir(parents=True, exist_ok=True)
    for destination, entries in outputs:
        with ZipFile(destination, "w", ZIP_DEFLATED) as archive:
            for source, archive_name in entries:
                # Fixed metadata makes the same source produce the same ZIP.
                info = ZipInfo(archive_name, date_time=(2020, 1, 1, 0, 0, 0))
                info.create_system = 3
                info.compress_type = ZIP_DEFLATED
                info.external_attr = 0o100644 << 16
                archive.writestr(info, source.read_bytes())
    return [path for path, _ in outputs]


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--out-dir", type=Path, default=ROOT / "dist")
    args = parser.parse_args()
    for path in build(ROOT, args.out_dir):
        print(path)


if __name__ == "__main__":
    main()

File: scripts/build-release.mjs
import { existsSync, mkdirSync, readFileSync, realpathSync, statSync, writeFileSync } from 'node:fs';
import { dirname, isAbsolute, join, relative, resolve, sep } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { parseArgs } from 'node:util';
import { zipSync } from 'fflate';

export const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');
export const SKILL_NAME = 'instagram-saved-ids';
export const SKILL_FILES = [
  'SKILL.md', 'agents/openai.yaml', 'references/browser-collection.md',
  'references/runtime.md', 'scripts/export_ids.js',
];

export function buildRelease(root = ROOT, output = join(root, 'dist')) {
  root = realpathSync(root);
  const manifest = JSON.parse(readFileSync(join(root, 'plugin.json'), 'utf8'));
  const legacy = JSON.parse(readFileSync(join(root, '.codex-plugin/plugin.json'), 'utf8'));
  const { name, version } = manifest;
  if (typeof name !== 'string' || !/^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(name)) throw new Error('Invalid package name');
  if (typeof version !== 'string' || !/^\d+\.\d+\.\d+$/.test(version)) throw new Error('Release version must use major.minor.patch');
  for (const field of ['name', 'version', 'description', 'license', 'repository']) {
    if (legacy[field] !== manifest[field]) throw new Error(`Compatibility manifest differs on ${field}`);
  }
  const skillEntries = SKILL_FILES.map(path => [`skills/${SKILL_NAME}/${path}`, `${SKILL_NAME}/${path}`]);
  const pluginEntries = ['plugin.json', '.codex-plugin/plugin.json', 'README.md'].map(path => [path, path]);
  pluginEntries.push(...SKILL_FILES.map(path => [`skills/${SKILL_NAME}/${path}`, `skills/${SKILL_NAME}/${path}`]));
  if (existsSync(join(root, 'LICENSE'))) {
    skillEntries.push(['LICENSE', `${SKILL_NAME}/LICENSE`]);
    pluginEntries.push(['LICENSE', 'LICENSE']);
  } else if (manifest.license) {
    throw new Error('Manifest declares a license but LICENSE is missing');
  }

  // Read only allowlisted regular files, and check all inputs before writing outputs.
  const inputs = new Map();
  for (const [source] of [...skillEntries, ...pluginEntries]) {
    const path = realpathSync(join(root, source));
    const inside = relative(root, path);
    if (!inside || isAbsolute(inside) || inside === '..' || inside.startsWith(`..${sep}`) || !statSync(path).isFile()) {
      throw new Error(`Missing or external package input: ${source}`);
    }
    inputs.set(source, readFileSync(path));
  }
  const archives = [
    [join(output, `${SKILL_NAME}-${version}.zip`), skillEntries],
    [join(output, `${name}-${version}.zip`), pluginEntries],
  ];
  mkdirSync(output, { recursive: true });
  for (const [destination, entries] of archives) {
    const files = Object.fromEntries(entries.map(([source, path]) => [path, inputs.get(source)]));
    writeFileSync(destination, zipSync(files, {
      level: 6, mtime: new Date(2020, 0, 1), os: 3, attrs: 0o100644 << 16,
    }));
  }
  return archives.map(([path]) => path);
}

if (process.argv[1] && pathToFileURL(resolve(process.argv[1])).href === import.meta.url) {
  const { values } = parseArgs({ options: { 'out-dir': { type: 'string', default: join(ROOT, 'dist') } } });
  for (const path of buildRelease(ROOT, resolve(values['out-dir']))) console.log(path);
}
